crates/data_components/src/arrow/write.rs

```diff
@@ -218,9 +235,84 @@ impl DisplayAs for MemSink {
 }
 
 impl MemSink {
-    fn new(batches: Vec<PartitionData>, overwrite: InsertOp) -> Self {
-        Self { batches, overwrite }
+    fn new(
+        batches: Vec<PartitionData>,
+        overwrite: InsertOp,
+        primary_key: Option<Vec<usize>>,
+    ) -> Self {
+        Self {
+            batches,
+            overwrite,
+            primary_key,
+        }
+    }
+}
+
+/// Create primary key values for a [`RecordBatch`]. For composite keys, values are concatenated with a delimiter '|'.
+fn extract_primary_keys_str(batch: &RecordBatch, pk_indices: &[usize]) -> Result<Vec<String>> {
```

Comment:
The order matters for the `pk_indices` - should we sort them to ensure they are always correct?